### DISCUSSION TITLE
Modify response from a hook?

### DISCUSSION THREAD

**Discussion by please-fill-out-this-field** [mojolicious/mojo#2184] [Q&A] [answered]
I need to modify the response of a certain controller action, let's say `foo#bar`, in a Mojolicious app I am not allowed to modify directly.

At first I thought of something like this:

```perl
$app->hook(before_render => sub {
  my ($c, $args) = @_;
  my $action = $c->match->stack->[-1]{action};

  return unless ref $c eq 'App::Controller::Foo' and $action eq 'bar';
  
  my $custom_data = 'hello world';
  $args->{json}{custom_data} = $custom_data;
});
```

However in my case `$custom_data` is a result of an asynchronous function returning `Mojo::Promise`:

```perl
$app->hook(before_render => sub {
  my ($c, $args) = @_;
  my $action = $c->match->stack->[-1]{action};

  return unless ref $c eq 'App::Controller::Foo' and $action eq 'bar';
  
  # need to wait for this before rendering stuff
  my $p = retrieve_custom_data->then(sub {
    my $custom_data = shift;
    $args->{json}{custom_data} = $custom_data;
  })->catch(\&handle_errors);
});
```

which means I need to delay rendering until `$p` is resolved.

It's not like I can just call `wait` on `$p` - that would have zero effect since the global Mojo::IOLoop singleton (aka `$p->ioloop`) is obviously already running at that point. Besides, I don't want to block the entire worker until it's resolved anyway.

Anything I can do here?

**Comment by Grinnz** [comment, marked as answer]
You wouldn't be able to cause the hook to wait for the result except by blocking. Only the around_action hook is powerful enough to dictate when the dispatch continues.

**Comment by jberger** [comment]
I'm sure there are legitimate needs for a non-blocking before_render hook, but in your case, checking the controller and action, couldn't you just append your hook code to your action? actions that return a promise are waited on by the loop
